lib/phoenix/live_dashboard/helpers/supervision_tree_helpers/supervision_tree.ex

```diff
@@ -0,0 +1,114 @@
+defmodule Phoenix.LiveDashboard.SupervisionTree do
+  # Construct supervision tree given an application master
+  @moduledoc false
+  alias Phoenix.LiveDashboard.{SystemInfo}
+
+  @node_width 100
+  @node_x_separation 50
+  @total_x_distance @node_width + @node_x_separation
+
+  def construct_tree(master) do
+    {pid, name} = SystemInfo.fetch_application_master_child(master)
+
+    child = get_nodes({name, pid, :supervisor, []}, 2, 0)
+    {:ok, [dictionary: dictionary]} = SystemInfo.fetch_process_info(pid, [:dictionary])
+
+    case Keyword.get(dictionary, :"$ancestors") do
+      [parent] ->
+        %{
+          pid: master,
+          name: name(master),
+          y: 1,
+          x: 0,
+          children: [
+            %{
+              pid: parent,
+              name: name(parent),
+              y: 1,
+              x: @total_x_distance,
+              children: child,
+              modifier: 0,
+              num: 0,
+              type: :subtree
+            }
+          ],
+          modifier: 0,
+          num: 0,
+          type: :subtree
+        }
+
+      _ ->
+        [child]
+    end
+  end
+
+  def get_nodes({_, :undefined, _, _}, _level, _num), do: []
+
+  def get_nodes({_, pid, :supervisor, _}, level, num) do
+    {:ok, [links: links]} = SystemInfo.fetch_process_info(pid, [:links])
+
+    {nodes, _} =
+      case Enum.count(links) do
+        1 ->
+          {[], :some}
+
+        _ ->
+          pid
+          |> SystemInfo.fetch_supervisor_children()
+          |> Kernel.++(Enum.filter(links, fn link -> is_port(link) end))
```

Comment:
```suggestion
          |> Kernel.++(Enum.filter(links, fn link -> is_port(link) end))
```

This line is only keeping ports and I think you want the opposite. My suggestion is to not worry about links on the initial version and we can add them later. :)

Reply:
I think I got how to make a tree like observer:

```elixir
defmodule AppTree do
  def app_tree(app) do
    master = :application_controller.get_master(app)
    {child, ^app} = :application_master.get_child(master)
    {nodes, _seen} = to_flat_node(:supervisor, child, master, %{master => true})
    {{:master, master, []}, nodes}
  end

  defp to_flat_node(type, pid, master, seen) do
    seen = Map.put(seen, pid, true)
    {children, seen} = children(type, pid, master, seen)
    {:registered_name, registered_name} = Process.info(pid, :registered_name)
    {[{{type, pid, registered_name}, children}], seen}
  end

  defp children(:supervisor, pid, master, seen) do
    case :supervisor.which_children(pid) do
      [] ->
        # If we don't have any child, it may be a supervisor bridge, so we also look at links.
        children(:worker, pid, master, seen)

      children ->
        Enum.flat_map_reduce(children, seen, fn {_id, child, type, _modules}, seen ->
          if is_pid(child) do
            to_flat_node(type, child, master, seen)
          else
            {[], seen}
          end
        end)
    end
  end

  defp children(:worker, pid, master, seen) do
    case Process.info(pid, :links) do
      {:links, children} ->
        Enum.flat_map_reduce(children, seen, fn child, seen ->
          if is_pid(child) and not Map.has_key?(seen, child) and
               Process.info(child, :group_leader) == {:group_leader, master} do
            to_flat_node(:worker, child, master, seen)
          else
            {[], seen}
          end
        end)

      _ ->
        {[], seen}
    end
  end
end

IO.inspect(AppTree.app_tree(:kernel))
```

This was trickier than I thought. Not only we need to get the links, as you did above, but we also need to keep all nodes we have seen so far to avoid infinite loops. My suggestion is to move the `app_tree` above to `SystemInfo` (instead of the three isolated functions) and then do a pass where you convert the app tree to a ReingoldTilford tree. :)

Reply:
It sounds good to me, I'm doing changes. :nerd_face:

Reply:
I made all changes, but theres some nodes that duplicates

![image](https://user-images.githubusercontent.com/25604253/82746306-df27df00-9d53-11ea-8674-95cd4521a134.png)

I'm trying to solve the problem. :monocle_face:

Reply:
Ohhhh, good catch. I think we need to do two passes on the tree, one building the supervisor tree and another building the links tree. I also noticed that both links and children are returned in reverse order, so we need to reverse those too. Here is the latest patch:

```elixir
defmodule AppTree do
  def app_tree(app) do
    master = :application_controller.get_master(app)
    {child, ^app} = :application_master.get_child(master)
    {children, seen} = sup_tree(child, %{master => true, child => true})
    {children, _seen} = links_tree(children, master, seen)
    {{:master, master, []}, [to_node(:supervisor, child, children)]}
  end

  defp sup_tree(pid, seen) do
    pid
    |> :supervisor.which_children()
    |> Enum.reverse()
    |> Enum.flat_map_reduce(seen, fn {_id, child, type, _modules}, seen ->
      if is_pid(child) do
        {children, seen} = if type == :worker, do: {[], seen}, else: sup_tree(child, seen)
        {[{type, child, children}], put_child(seen, child)}
      else
        {[], seen}
      end
    end)
  end

  defp links_tree(nodes, master, seen) do
    Enum.map_reduce(nodes, seen, fn {type, pid, children}, seen ->
      {children, seen} =
        if children == [], do: links_children(type, pid, master, seen), else: {children, seen}

      {children, seen} = links_tree(children, master, seen)
      {to_node(type, pid, children), seen}
    end)
  end

  defp links_children(parent_type, pid, master, seen) do
    # If the parent type is a supervisor and we have no children,
    # then this may be a supervisor bridge, so we tag its children
    # as workers, otherwise they are links.
    type = if parent_type == :supervisor, do: :worker, else: :link

    case Process.info(pid, :links) do
      {:links, children} ->
        children
        |> Enum.reverse()
        |> Enum.flat_map_reduce(seen, fn child, seen ->
          if is_pid(child) and not has_child?(seen, child) and has_leader?(child, master) do
            {[{type, child, []}], put_child(seen, child)}
          else
            {[], seen}
          end
        end)

      _ ->
        {[], seen}
    end
  end

  defp to_node(type, pid, children) do
    {:registered_name, registered_name} = Process.info(pid, :registered_name)
    {{type, pid, registered_name}, children}
  end

  defp has_child?(seen, child), do: Map.has_key?(seen, child)
  defp put_child(seen, child), do: Map.put(seen, child, true)
  defp has_leader?(pid, gl), do: Process.info(pid, :group_leader) == {:group_leader, gl}
end
```